src/netlink.c

```diff
@@ -383,10 +405,190 @@ static int global_create(lua_State *L) {
         nl->fd = sock;
         nl->type = SOCK_RAW;
         timeout_init(&nl->tm, -1, -1);
+        nl->nl_family_id = 0;
         return 1;
     }
 
     lua_pushnil(L);
     lua_pushstring(L, socket_strerror(err));
     return 2;
 }
+
+/*-------------------------------------------------------------------------*\
+ * Generic netlink socket functions.
+ *
+ * resolves netlink family id.
+\*-------------------------------------------------------------------------*/
+static int resolve_nl_family_id(lua_State *L) {
+    char family_name[32] = {0};
+    struct sockaddr_nl nl_address;
+    size_t sent = 0;
+    int err;
+    struct nlattr *nl_na;
+    int nl_family_id = -1;
+
+    sprintf(family_name, "NFLUA");
+    p_netlink nl = (p_netlink)auxiliar_checkclass(L, "netlink{unconnected}", 1);
+    p_timeout tm = &nl->tm;
+
+    memset(nl->nlgb, 0, sizeof(struct nlgenmsgbuf));
+
+    nl->nlgb->n.nlmsg_type = GENL_ID_CTRL;
+    nl->nlgb->n.nlmsg_flags = NLM_F_REQUEST;
+    nl->nlgb->n.nlmsg_seq = 0;
+    nl->nlgb->n.nlmsg_pid = luaL_checkinteger(L, 3);
+    nl->nlgb->n.nlmsg_len = NLMSG_LENGTH(GENL_HDRLEN);
+    nl->nlgb->g.cmd = CTRL_CMD_GETFAMILY;
+    nl->nlgb->g.version = 0x1;
+
+    nl_na = (struct nlattr *) GENLMSG_DATA(nl->nlgb);
+    nl_na->nla_type = CTRL_ATTR_FAMILY_NAME;
+    nl_na->nla_len = strlen(family_name) + 1 + NLA_HDRLEN;
+    strcpy(NLA_DATA(nl_na), family_name);
+    nl->nlgb->n.nlmsg_len += NLMSG_ALIGN(nl_na->nla_len);
+
+    memset(&nl_address, 0, sizeof(nl_address));
+    nl_address.nl_family = AF_NETLINK;
+    nl_address.nl_pid = luaL_checkinteger(L, 3);
+
+    socklen_t len = sizeof(nl_address);
+    timeout_markstart(tm);
+    err = socket_sendto(&nl->fd, (char *) nl->nlgb, nl->nlgb->n.nlmsg_len, &sent,
+            (SA *)&nl_address, len, tm);
+    if (err != IO_DONE) {
+        goto out;
+    }
+
+    memset(nl->nlgb, 0, sizeof(struct nlgenmsgbuf));
+    timeout_markstart(tm);
+    err = socket_recvfrom(&nl->fd, (char *)nl->nlgb, NLMSG_SPACE(MAX_PAYLOAD) + GENL_HDRLEN, &sent,
+            (SA *)&nl_address, &len, tm);
+    if (err != IO_DONE && err != IO_CLOSED) {
+        goto out;
+    }
+
+    if (!NLMSG_OK(&nl->nlgb->n, sent)) {
+        goto out;
+    }
+
+    if (nl->nlgb->n.nlmsg_type == NLMSG_ERROR) {
+        goto out;
+    }
+
+    nl_na = (struct nlattr *) GENLMSG_DATA(nl->nlgb);
+    nl_na = (struct nlattr *) ((char *) nl_na + NLA_ALIGN(nl_na->nla_len));
```

Comment:
This skips over the first attribute, right? What is that attribute? Should it be checked for being as expected? Is it guaranteed that we will receive `CTRL_ATTR_FAMILY_ID` second, or should we loop over all attributes to see which one it is?

Reply:
In kernel there is:
```
       if (nla_put_string(skb, CTRL_ATTR_FAMILY_NAME, family->name) ||
            nla_put_u16(skb, CTRL_ATTR_FAMILY_ID, family->id) ||
            nla_put_u32(skb, CTRL_ATTR_VERSION, family->version) ||
            nla_put_u32(skb, CTRL_ATTR_HDRSIZE, family->hdrsize) ||
            nla_put_u32(skb, CTRL_ATTR_MAXATTR, family->maxattr))
                goto nla_put_failure;
```
So first would be family name and ID is always second.

Reply:
I still feel uneasy relying on it, even though I would expect the kernel to remain compatible when it comes to this sort of thing. Maybe at least add a comment to this line about skipping over the family name?

Reply:
Sure future kernels can change. I'll add the comment for now, we can improve later.